rust/connlib/tunnel/src/peer.rs

```diff
@@ -335,6 +535,87 @@ pub struct ClientOnGateway {
     ipv6: Ipv6Addr,
     resources: HashMap<ResourceId, Vec<ResourceOnGateway>>,
     filters: IpNetworkTable<FilterEngine>,
+    permanent_translations: HashMap<IpAddr, (IpAddr, ResourceId, DomainName, Instant, bool)>,
+    nat_table: NatTable,
+    buffered_events: VecDeque<GatewayEvent>,
+}
+
+#[derive(Default, Debug)]
+struct NatTable {
+    table: BiMap<(Protocol, IpAddr), (Protocol, IpAddr)>,
+    last_seen: HashMap<(Protocol, IpAddr), Instant>,
+}
+
+impl NatTable {
+    fn expire(&mut self, now: Instant) {
+        let mut removed = Vec::new();
+        for (r, e) in self.last_seen.iter() {
+            if now.duration_since(*e) >= Duration::from_secs(60) {
+                self.table.remove_by_right(r);
+                removed.push(*r);
+            }
+        }
+
+        for r in removed {
+            self.last_seen.remove(&r);
+        }
+    }
+
+    fn translate(
+        &mut self,
+        outgoing_pkt: &IpPacket,
+        real_address: IpAddr,
+        now: Instant,
+    ) -> Option<(Protocol, IpAddr)> {
+        let source_protocol = outgoing_pkt.source_protocol()?;
+        if let Some(res) = self
+            .table
+            .get_by_left(&(source_protocol, outgoing_pkt.destination()))
+        {
+            if res.1 == real_address {
+                self.last_seen.insert(*res, now);
+                return Some(*res);
+            }
+        }
+
+        let mut occupied_ports = self
+            .table
+            .iter()
+            .filter(|(_, (proto, ip))| *ip == real_address && proto.same_type(&source_protocol))
+            .map(|(_, (proto, _))| proto.value())
+            .sorted_unstable();
+
+        for p in 1.. {
+            if !occupied_ports.contains(&p) {
+                let proxy_protocol = source_protocol.with_value(p);
+                self.table.insert(
+                    (source_protocol, outgoing_pkt.destination()),
+                    (proxy_protocol, real_address),
+                );
+                self.last_seen.insert((proxy_protocol, real_address), now);
+                return Some((proxy_protocol, real_address));
+            }
+        }
```

Comment:
if it is a packet going from the same source port and destination ip within the session window (30 seconds) it keeps using the same port.

If we want to make it more stable we can just use the same source port if it's available and the increment from there? I think that's better than a hash function.

Reply:
That is unlikely to be true though, right? Likely, many clients will be using the same apps on the same gateway so only the first one will get the same port. A hash function that takes into account the source IP would be deterministic per client but different for each user.

Reply:
but this nat table is per-client anyways

Reply:
Ah right, I forgot about that. Yeah then same port is good enough!

Reply:
now I realize that the current implementation is not only per-client but consider as occupied ports if it has the same real ip, so if we use same port it will only collide in a very small number of cases :)